nogotofail/mitm/connection/handlers/data/ssl.py

```diff
@@ -172,3 +175,66 @@ def on_ssl(self, client_hello):
                 self.log(logging.ERROR,
                         "Client enabled SSLv3 protocol without TLS_FALLBACK_SCSV")
             self.log_attack_event(data="SSLv3")
+
+
+@handler.passive(handlers)
+class NoForwardSecrecy(DataHandler):
+    name = "noforwardsecrecy"
+    description = (
+        "Detects selected server cipher suites which don't support "
+        "Diffie-Hellman key exchange (DHE or ECDHE) i.e. in "
+        "SERVER_HELLO response")
+
+    buffer = ""
+    ssl = False
+
+    def on_ssl(self, client_hello):
+        self.ssl = True
+        self.client_session_id = client_hello.session_id
+        return False
+
+    def on_response(self, response):
+        if not self.ssl:
+            return response
+        response = self.buffer + response
+        self.buffer = ""
+        try:
+            record, remaining = tls.parse_tls(response)
+            for i, message in enumerate(record.messages):
+                # Check for Server Hello message
+                if (isinstance(message, tls.types.HandshakeMessage) and
+                        message.type == HandshakeMessage.TYPE.SERVER_HELLO):
+                    server_hello = message.obj
+                    selected_cipher = str(server_hello.cipher)
+                    _connection = self.connection
+                    destination = _connection.hostname if \
+                        _connection.hostname else _connection.server_addr
+                    debug_message = ["Selected cipher \"", selected_cipher,
+                            "\" for connection to \"", destination, "\""]
+                    self.log(logging.DEBUG, "".join(debug_message))
+                    """ Check if Ephemeral Diffie-Hellman key exchange is
+                        used in selected cipher """
+                    fs_key_strings = ["DHE", "ECDHE"]
+                    if not [fs_string for fs_string in fs_key_strings
+                            if fs_string in selected_cipher]:
+                        error_message = \
+                            ["Cipher suite key exhange technqiue doesn't ",
+                             "support forward secrecy. ",
+                             "Cipher suite - [", selected_cipher, "]"]
+                        self.log(logging.WARNING, "".join(error_message))
```

Comment:
Is this rare enough for WARNING? It might be worth dropping it to INFO if it happens very often (to avoid spamming).